examples/raytracing/example_raytracing.cpp

```diff
@@ -119,83 +231,182 @@ int main(int argc, char *argv[])
-  timer.reset();
-  ArborX::BVH<MemorySpace> bvh{exec_space,
-                               SpheresToBoxes<MemorySpace>{spheres}};
-
-  Kokkos::View<float *, MemorySpace> accumulator("accumulator", num_rays);
-  bvh.query(exec_space, Rays<MemorySpace>{rays},
-            AccumRaySphereInterDist<MemorySpace>{spheres, accumulator});
-  exec_space.fence();
-  auto time = timer.seconds();
+  // OrderedIntersects-based approach
+  Kokkos::View<float *, MemorySpace> energy_ordered_intersects;
+  {
+    Kokkos::Profiling::pushRegion("Example::ordered_intersects_approach");
+    Kokkos::View<float *, MemorySpace> ray_energy(
+        Kokkos::view_alloc("Example::ray_energy", Kokkos::WithoutInitializing),
+        num_rays * num_boxes);
+    Kokkos::deep_copy(ray_energy, (total_energy * dx * dy * dz) / num_rays);
+    energy_ordered_intersects = Kokkos::View<float *, MemorySpace>(
+        "Example::energy_ordered_intersects", num_boxes);
+
+    bvh.query(exec_space, OrderedIntersectsBased::Rays<MemorySpace>{rays},
+              OrderedIntersectsBased::DepositEnergy<MemorySpace>{
+                  boxes, ray_energy, energy_ordered_intersects});
+    Kokkos::Profiling::popRegion();
+  }
 
-  auto accumulator_avg =
-      ArborX::accumulate(exec_space, accumulator, 0.f) / num_rays;
+  // Intersects-based approach
+  Kokkos::View<float *, MemorySpace> energy_intersects;
+  {
+    Kokkos::Profiling::pushRegion("Example::intersects_approach");
+    Kokkos::View<IntersectsBased::IntersectedCell *> values("Example::values",
+                                                            0);
+    Kokkos::View<int *> offsets("Example::offsets", 0);
+    bvh.query(
+        exec_space, IntersectsBased::Rays<MemorySpace>{rays},
+        IntersectsBased::AccumulateRaySphereIntersections<MemorySpace>{boxes},
+        values, offsets);
+
+    Kokkos::View<IntersectsBased::IntersectedCellForSorting *, MemorySpace>
+        sort_array(Kokkos::view_alloc(exec_space, Kokkos::WithoutInitializing,
+                                      "Example::sort_array"),
+                   values.size());
+    Kokkos::parallel_for(
+        "Example::copy sort_array",
+        Kokkos::RangePolicy<ExecutionSpace>(exec_space, 0, values.size()),
+        KOKKOS_LAMBDA(int i) { sort_array(i) = values(i); });
+    Kokkos::Profiling::pushRegion("Example::sorting");
+#if defined(KOKKOS_ENABLE_CUDA) || defined(KOKKOS_ENABLE_HIP) ||               \
+    defined(KOKKOS_ENABLE_SYCL)
+    auto permutation = ArborX::Details::sortObjects(exec_space, sort_array);
+#else
+    Kokkos::View<int *> permutation(
+        Kokkos::view_alloc("Example::permutation", Kokkos::WithoutInitializing),
+        sort_array.size());
+    std::iota(permutation.data(), permutation.data() + sort_array.size(), 0);
+    std::sort(permutation.data(), permutation.data() + sort_array.size(),
+              [&](int const &a, int const &b) {
+                return (sort_array(a) < sort_array(b));
+              });
+#endif
```

Comment:
How much does this impact performance?  We are showcasing something we don't want user to do, that is reaching into the `Details::` namespace.
The region label should read "sort[ing] by key".
If the impact on performance is significant I would be ok marking this as a FIXME and using it as a excuse to go ahead and add an abstraction for it in `KokkosExt::` (which we still don't really want users to reach into)

Reply:
Sorting on the host for the benchmark I have been using throughout is around 100x more expensive.